clang/test/Sema/warn-jump-bypasses-init.c

```diff
@@ -0,0 +1,31 @@
+// RUN: %clang_cc1 -fsyntax-only -verify=c,both -Wjump-bypasses-init %s
+// RUN: %clang_cc1 -fsyntax-only -verify=c,both -Wc++-compat %s
+// RUN: %clang_cc1 -fsyntax-only -verify=good %s
+// RUN: %clang_cc1 -fsyntax-only -verify=cxx,both -x c++ %s
+// good-no-diagnostics
+
+void goto_func_1(void) {
```

Comment:
How about:

```c
void f(void) {          
  ({ 
    goto ouch;  
    int i = 12;
  });

ouch:
  ;
}
```

Reply:
Also:

```c
void f(void) {          

 for (int i=({ goto ouch; int x = 10; x;});i<0;++i)

ouch:
  ;
}
```

Reply:
Those are both valid in C++, so I can add them as negative tests: https://godbolt.org/z/G86YqoToK

Reply:
We currently accept those in C++ the same as GCC: https://godbolt.org/z/f4ze65vWT but I'm not convinced that's a good behavior. I think those *should* be rejected in C++. And I think this patch actually makes us start to reject them, so good call!

Reply:
These are fun examples. It turns out that this code is accepted in both C and C++ despite jumping over the initialization, in both Clang and GCC. I'm not convinced that's a good design, but I don't think this patch changes the behavior either, and so I don't think we should diagnose. However, perhaps we should in a follow-up, so I've filed: https://github.com/llvm/llvm-project/issues/138257 and added these as negative tests for this PR.